subprojects/dependency-management/src/main/java/org/gradle/internal/resource/cached/DefaultCachedExternalResourceIndex.java

```diff
@@ -30,7 +31,7 @@
     private final BuildCommencedTimeProvider timeProvider;
 
     public DefaultCachedExternalResourceIndex(String persistentCacheFile, Class<K> keyType, BuildCommencedTimeProvider timeProvider, CacheLockingManager cacheLockingManager) {
-        super(persistentCacheFile, new DefaultSerializer<K>(keyType.getClassLoader()), new DefaultSerializer<CachedExternalResource>(CachedExternalResource.class.getClassLoader()), cacheLockingManager);
+        super(persistentCacheFile, new WellKnownTypesSerializer<K>(keyType.getClassLoader()), new DefaultSerializer<CachedExternalResource>(CachedExternalResource.class.getClassLoader()), cacheLockingManager);
```

Comment:
The serializers should be provided by the caller, which has the knowledge to select the most efficient serializer.

We should use a more efficient serializer for the value as well as the key.